File: src/test/java/org/snakeyaml/engine/usecases/untrusted/ByteLimitTest.java
/*
 * Copyright (c) 2018, SnakeYAML
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 */
package org.snakeyaml.engine.usecases.untrusted;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.fail;

import java.util.Iterator;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.snakeyaml.engine.v2.api.Load;
import org.snakeyaml.engine.v2.api.LoadSettings;

public class ByteLimitTest {

  @Test
  @DisplayName("Limit a single document")
  public void testSetCodePointLimit() {
    LoadSettings settings = LoadSettings.builder().setCodePointLimit(15).build();
    Load load = new Load(settings);
    try {
      load.loadFromString("12345678901234567890");
      fail("Long input should not be accepted");
    } catch (Exception e) {
      assertEquals("The incoming YAML document exceeds the limit: 15 code points.", e.getMessage());
    }
  }

  @Test
  public void testLoadAll553() {
    LoadSettings settings = LoadSettings.builder().setCodePointLimit(15).build();
    Load load = new Load(settings);
    try {
      Iterator<Object> iter = load.loadAllFromString("12345678901234567890").iterator();
      iter.next();
      fail("Long input should not be accepted for loadAll");
    } catch (Exception e) {
      assertEquals("The incoming YAML document exceeds the limit: 15 code points.", e.getMessage());
    }
  }

  @Test
  public void testLoadManyDocuments() {
    LoadSettings settings = LoadSettings.builder().setCodePointLimit(8).build();
    Load load = new Load(settings);
    Iterator<Object> iter = load.loadAllFromString("---\nfoo\n---\nbar\n---\nyep").iterator();
    assertEquals("foo", iter.next());
    assertEquals("bar", iter.next());
    assertEquals("yep", iter.next());
    assertFalse(iter.hasNext());
  }
}

File: src/commonTest/kotlin/it/krzeminski/snakeyaml/engine/kmp/usecases/untrusted/ByteLimitTest.kt
package it.krzeminski.snakeyaml.engine.kmp.usecases.untrusted

import io.kotest.assertions.throwables.shouldThrow
import io.kotest.core.spec.style.FunSpec
import io.kotest.matchers.shouldBe
import it.krzeminski.snakeyaml.engine.kmp.api.Load
import it.krzeminski.snakeyaml.engine.kmp.api.LoadSettings

class ByteLimitTest : FunSpec({
    test("limit a single document") {
        val settings = LoadSettings.builder().setCodePointLimit(15).build()
        val load = Load(settings)
        shouldThrow<Exception> {
            load.loadOne("12345678901234567890")
        }.also { exception ->
            exception.message shouldBe "The incoming YAML document exceeds the limit: 15 code points."
        }
    }

    test("load all 553") {
        val settings = LoadSettings.builder().setCodePointLimit(15).build()
        val load = Load(settings)
        shouldThrow<Exception> {
            val iter = load.loadAll("12345678901234567890").iterator()
            iter.next()
        }.also { exception ->
            exception.message shouldBe "The incoming YAML document exceeds the limit: 15 code points."
        }
    }

    test("load many documents") {
        val settings = LoadSettings.builder().setCodePointLimit(8).build()
        val load = Load(settings)
        val iter = load.loadAll("---\nfoo\n---\nbar\n---\nyep").iterator()
        iter.next() shouldBe "foo"
        iter.next() shouldBe "bar"
        iter.next() shouldBe "yep"
        iter.hasNext() shouldBe false
    }
})
